lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Random.hs

```diff
@@ -268,3 +280,90 @@ instance CompareDiscovery (RndState n) where
 
 instance KnownAddresses (RndState n) where
     knownAddresses s = Map.elems (addresses s)
+
+--------------------------------------------------------------------------------
+--
+-- RndAnyState
+--
+-- For benchmarking and testing arbitrary large random wallets.
+
+-- | An "unsound" alternative that can be used for benchmarking and stress
+-- testing. It re-uses the same underlying structure as the `RndState` but
+-- it discover addresses based on an arbitrary ratio instead of decrypting the
+-- derivation path.
+--
+-- The proportion is stored as a type-level parameter so that we don't have to
+-- alter the database schema to store it. It simply exists and depends on the
+-- caller creating the wallet to define it.
+newtype RndAnyState (network :: NetworkDiscriminant) (p :: Nat) = RndAnyState
+    { innerState :: RndState network
+    } deriving (Generic, Show)
+
+instance NFData (RndAnyState n p)
+
+-- | Initialize the HD random address discovery state from a root key and RNG
+-- seed.
+--
+-- The type parameter is expected to be a ratio (between 0 and 100) of addresses
+-- we ought to simply recognize as ours. So, giving @5 means that 5% of the
+-- entire address space of the network will be considered ours, picked randomly.
+mkRndAnyState
+    :: forall (p :: Nat) n. ()
+    => ByronKey 'RootK XPrv
+    -> Int
+    -> RndAnyState n p
+mkRndAnyState key seed = RndAnyState
+    { innerState = RndState
+        { hdPassphrase = payloadPassphrase key
+        , accountIndex = minBound
+        , addresses = mempty
+        , pendingAddresses = mempty
+        , gen = mkStdGen seed
+        }
+    }
+
+instance KnownNat p => IsOurs (RndAnyState n p) Address where
+    isOurs addr@(Address bytes) st@(RndAnyState inner) =
+        case isOurs addr inner of
+            (True, inner') ->
+                (True, RndAnyState inner')
+
+            (False, _) | crc32 bytes < p ->
+                let
+                    (path, gen') = findUnusedPath
+                        (gen inner) (accountIndex inner) (unavailablePaths inner)
```

Comment:
Couldn't this extra `findUnusedPath` call, compared to the inner state implementation, affect the results?

Reply:
A bit yes. We don't benchmark this function independently but I believe that even on large wallets with 100K+ addresses, finding an unused path in a space of 2^31 is quite fast.

Reply:
So I'd expect it to be quite negligible.